### PR TITLE
fix(maxInactive): Refactor MaxInactiveInterceptor to MaxInactiveFilter

### LINKED ISSUE #1742: Fix check for session max inactive
**Is your feature request related to a problem? Please describe.**
For installations that use a separate front-end, there needs to be a way to "touch" the user session to avoid timing out.

**Describe the solution you'd like**
Implement this as an API call that checks the user's currently configured session duration and return the value as JSON.

**Describe alternatives you've considered**
A more ambitious alternative is to have all (most?) API endpoints check if there is a session cookie and update session. "Touching a session" is not currently implemented in the related Tomcat spec.

**Additional context**
Again, this will help those front-ends that heavily leverage APIs

### PR DESCRIPTION
**Pull request by bjagg**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]
-   [x] tests are included

##### Description of change

Resolves #1742 

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java [hunk 1]
```diff
@@ -1,78 +1,109 @@
...
 
-public class MaxInactiveInterceptor extends HandlerInterceptorAdapter {
+/**
+ * This is a refactor to correct the application of the session timeout (aka MaxInactive time) to
+ * make sure it is applied consistently. The {@code HandlerInterceptorAdapter} version was not
+ * applying correctly to JSON and other corner cases.
+ *
+ * @since 5.7.1
+ */
+@Slf4j
```
Comment (ChristianMurphy) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java]: Lombok annotations :heart_eyes:

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/RequireValidSessionFilter.java [hunk 2]
```diff
@@ -16,15 +16,20 @@
...
 import javax.servlet.http.HttpSession;
+import lombok.extern.slf4j.Slf4j;
 import org.apereo.portal.security.mvc.LoginController;
 import org.springframework.web.filter.OncePerRequestFilter;
 
 /** Redirects the user to the Login servlet if they don't already have a session. */
+@Slf4j
 public class RequireValidSessionFilter extends OncePerRequestFilter {
+    //  private static final org.slf4j.Logger log =
+    // org.slf4j.LoggerFactory.getLogger(RequireValidSessionFilter.class);
```
Comment (ChristianMurphy) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/RequireValidSessionFilter.java]: could this be removed now that the class is annotated?

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java [hunk 3]
```diff
@@ -1,78 +1,109 @@
...
     @Override
-    public boolean preHandle(
-            HttpServletRequest request, HttpServletResponse response, Object handler)
-            throws Exception {
+    public void destroy() {}
+
+    @Override
+    public void doFilter(ServletRequest req, ServletResponse resp, FilterChain filterChain)
+            throws IOException, ServletException {
+        log.info("Entering MaxInactiveFilter");
```
Comment (jonathanmtran) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java]: Should this and line 69 be `log.debug`?

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/ClassicMaxInactiveStrategy.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        assert person != null;
+
+        IAuthorizationPrincipal principal =
+                authorizationService.newPrincipal(
+                        (String) person.getAttribute(IPerson.USERNAME), IPerson.class);
+        IPermission[] permissions =
+                authorizationService.getAllPermissionsForPrincipal(
+                        principal, IPermission.PORTAL_SYSTEM, MAX_INACTIVE_ATTR, null);
+
+        assert (permissions != null);
```
Comment (cbeach47) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/ClassicMaxInactiveStrategy.java]: Consider removing the parenthesis.   Noting line 35.

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        final HttpServletRequest request = (HttpServletRequest) req;
+        checkMaxInactive(request);
+
+        log.info("Continuing on to other filters from MaxInactiveFilter");
```
Comment (cbeach47) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java]: Probably want this as `log.debug` as well.

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/RequireValidSessionFilter.java [hunk 6]
```diff
@@ -39,7 +42,14 @@ protected boolean shouldNotFilter(HttpServletRequest request) {
         final HttpSession session = request.getSession(false);
         if (session != null && !session.isNew()) {
             // Session exists and is not new, don't bother filtering
+            log.debug("User {} has a session: {}", request.getRemoteUser(), session.getId());
+            log.debug("Max inactive interval: {}", session.getMaxInactiveInterval());
+            final Instant ctime = Instant.ofEpochMilli(session.getCreationTime());
+            final Instant atime = Instant.ofEpochMilli(session.getLastAccessedTime());
+            log.debug("Session creation time: {}, last access time: {}", ctime, atime);
```
Comment (cbeach47) [inline uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/RequireValidSessionFilter.java]: Consider wrapping lines 47 - 49 in a `log.isDebugEnabled() conditional`.

File: uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/MaxInactiveFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+        filter.doFilter(req, resp, chain);
+
+        verifyZeroInteractions(personManager);
+        verifyZeroInteractions(maxInactiveStrategy);
+        verifyZeroInteractions(resp);
+        verify(req, only()).getSession(false);
+        verify(chain, only()).doFilter(req, resp);
+    }
+}
```
Comment (cbeach47) [inline uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/MaxInactiveFilterTest.java]: :+1:  nice to see all the test coverage.

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java (hunk 3 above)
Comment (bjagg) [reply uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java]: :+1:

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/ClassicMaxInactiveStrategy.java (hunk 4 above)
Comment (bjagg) [reply uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/ClassicMaxInactiveStrategy.java]: :+1:

File: uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java (hunk 5 above)
Comment (bjagg) [reply uPortal-security/uPortal-security-mvc/src/main/java/org/apereo/portal/url/MaxInactiveFilter.java]: :+1:

File: uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/MaxInactiveFilterTest.java (hunk 7 above)
Comment (bjagg) [reply uPortal-security/uPortal-security-mvc/src/test/java/org/apereo/portal/url/MaxInactiveFilterTest.java]: :-)

**Pull request merged**